### ISSUE TITLE
pip install on Windows causes "failed to create process"

### ISSUE THREAD

**Issue by evakili** [conan-io/conan#298]
When I use `pip install conan` to install **conan**, it will be installed successfully but then runinng `conan` will be failed by this error:

```
failed to create process
```

My config is:
Windows 10 64bit
Python 3.5.1 64bit

**Comment by memsharded** [comment]
Could you please copy here the output of running "python" and "pip --version"? 
I think it is possible that the Python distribution could be 32bits, would you mind trying with a 32bits distro? Also please consider that the support for Python3 has been introduced recently, so it should be a bit less mature than for Python27. 
Have you used virtualenvs? I strongly recommend virtualenvwrapper (easily install in win with virtualenvwrapper-win) for everything python related.

**Comment by evakili** [comment]
Python 3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:54:25) [MSC v.1900 64 bit (AMD64)] on win32
    
    pip 8.1.2 from c:\...\python3\tools\lib\site-packages (python 3.5)

I'm not using `virtualenv`.
The **conan** works fine by installing via Windows installer you provided.

**Comment by memsharded** [comment]
Glad that it works with the installer. The question is, are you using Python 3.5 for something else? Would you mind installing instead the x86 version? https://www.python.org/ftp/python/3.5.1/python-3.5.1.exe

If you are just fine running with the installer version, then this issue could be closed. Thank you!

**Comment by evakili** [comment]
I have no problem with anything other than `pip install` with Python 3.5-64. I think you should not close this issue, although I understand that you haven't plan to fully support Python 3 for now.

**Comment by memsharded** [comment]
Yes we plan to support Python3, we will improve it, accept pull rquests related to it... it is just that it was recently introduced and it might be slightly less mature. This seems not to be a Python3 issue, but a Python3 64bits - 32bits issue. There are other users using Python3 from pip successfully, so my guess was the python 32/64. Ok, lets keep this open, it is not an urgent issue, as you can work from the installer, we will have a look when possible. Thanks again!

**Comment by ethanhs** [comment]
Just tested with Python 3.5.1 x86 on Windows 10, and `pip install conan` installed fine, so I am pretty sure the issue is with x86-64 Python 3.

**Comment by KindDragon** [comment]
Same issue with:
```
> python
Python 3.5.2 (v3.5.2:4def2a2901a5, Jun 25 2016, 22:01:18) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
```

**Comment by IntegralLee** [comment]
I ran into this issue also with both the 32-bit and 64-bit versions of Python.  When installing Python, I used the defaults, so in both cases, Python was installed into either `C:\Program Files` or `C:\Program Files (x64)`.  However, everything works properly if I install Python to `C:\Python35`.  Even with the 64-bit version.

I suspect the issue is actually the space in the path as covered in [https://github.com/pypa/pip/issues/2783](https://github.com/pypa/pip/issues/2783).

**Comment by memsharded** [comment]
Probably, in fact I always install python to C:/PythonXX in windows.

I am trying to reproduce the issue with 32 bits python3, without success:

```bash
$ which python
/c/Users/memsharded/Envs/conanpip3/Scripts/python
$ python
Python 3.5.1 (v3.5.1:37a07cee5969, Dec  6 2015, 01:38:48) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
```
Conan installs and runs just fine. I am using virtualenvs, and the path for python is also without spaces.

**Comment by KindDragon** [comment]
> I am trying to reproduce the issue with 32 bits python3 (as reported by @KindDragon to fail), without success:

Still can reproduce:
```
> pip install conan
> python
Python 3.5.2 (v3.5.2:4def2a2901a5, Jun 25 2016, 22:01:18) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
> where python
c:\Program Files (x86)\Python35-32\python.exe
> where conan
c:\Program Files (x86)\Python35-32\Scripts\conan.exe
```

"c:\Program Files (x86)\Python35-32\Lib\site-packages\conans\conan.py" works fine

**Comment by KindDragon** [comment]
Same issue with 64-bit Python:
```
$ python
Python 3.5.2 (v3.5.2:4def2a2901a5, Jun 25 2016, 22:18:55) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
```

**Comment by KindDragon** [comment]
`pip install conan` show error:
```
> pip install conan
Collecting conan
  Using cached conan-0.11.0.tar.gz
Collecting PyJWT<1.5.0,>=1.4.0 (from conan)
  Using cached PyJWT-1.4.2-py2.py3-none-any.whl
Collecting cfgparse<1.4.0,>=1.3 (from conan)
  Using cached cfgparse-1.3.zip
Collecting requests<2.8.0,>=2.7.0 (from conan)
  Using cached requests-2.7.0-py2.py3-none-any.whl
Collecting colorama<0.4.0,>=0.3.3 (from conan)
  Using cached colorama-0.3.7-py2.py3-none-any.whl
Collecting passlib<1.7.0,>=1.6.5 (from conan)
  Using cached passlib-1.6.5-py2.py3-none-any.whl
Collecting boto<2.39.0,>=2.38.0 (from conan)
  Using cached boto-2.38.0-py2.py3-none-any.whl
Collecting PyYAML<3.12.0,>=3.11 (from conan)
  Using cached PyYAML-3.11.zip
Collecting patch==1.16 (from conan)
  Using cached patch-1.16.zip
Collecting fasteners>=0.14.1 (from conan)
  Using cached fasteners-0.14.1-py2.py3-none-any.whl
Collecting six>=1.10.0 (from conan)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting bottle<0.13,>=0.12.8 (from conan)
  Using cached bottle-0.12.9.tar.gz
Collecting monotonic>=0.1 (from fasteners>=0.14.1->conan)
  Using cached monotonic-1.2-py2.py3-none-any.whl
Installing collected packages: PyJWT, cfgparse, requests, colorama, passlib, boto, PyYAML, patch, six, monotonic, fasteners, bottle, conan
  Running setup.py install for cfgparse ... done
  Running setup.py install for PyYAML ... done
  Running setup.py install for patch ... done
  Running setup.py install for bottle ... error
    Complete output from command c:\users\аркадий\appdata\local\programs\python\python35\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\E51C~1\\AppData\\Local\\Temp\\pip-build-dn75ia2v\\bottle\\setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record C:\Users\E51C~1\AppData\Local\Temp\pip-_fge1xwb-record\install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build\lib
    copying bottle.py -> build\lib
    running build_scripts
    creating build\scripts-3.5
    copying and adjusting bottle.py -> build\scripts-3.5
    Traceback (most recent call last):
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\command\build_scripts.py", line 114, in copy_scripts
        shebang.decode('utf-8')
    UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe0 in position 11: invalid continuation byte

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\E51C~1\AppData\Local\Temp\pip-build-dn75ia2v\bottle\setup.py", line 38, in <module>
        'Programming Language :: Python :: 3.3',
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\site-packages\setuptools\command\install.py", line 61, in run
        return orig.install.run(self)
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\command\install.py", line 539, in run
        self.run_command('build')
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\command\build.py", line 135, in run
        self.run_command(cmd_name)
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\command\build_scripts.py", line 50, in run
        self.copy_scripts()
      File "c:\users\рЁърфшщ\appdata\local\programs\python\python35\lib\distutils\command\build_scripts.py", line 118, in copy_scripts
        "from utf-8".format(shebang))
    ValueError: The shebang (b'#!c:\\users\\\xe0\xf0\xea\xe0\xe4\xe8\xe9\\appdata\\local\\programs\\python\\python35\\python.exe\n') is not decodable from utf-8

    ----------------------------------------
Command "c:\users\аркадий\appdata\local\programs\python\python35\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\E51C~1\\AppData\\Local\\Temp\\pip-build-dn75ia2v\\bottle\\setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record C:\Users\E51C~1\AppData\Local\Temp\pip-_fge1xwb-record\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\E51C~1\AppData\Local\Temp\pip-build-dn75ia2v\bottle\
You are using pip version 8.1.1, however version 8.1.2 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.
```

**Comment by KindDragon** [comment]
`pip install conan` work with Python 2.7.11 BTW

**Comment by KindDragon** [comment]
Probably related to https://mail.python.org/pipermail/python-win32/2015-November/013566.html
Same results with 0.12

**Comment by KindDragon** [comment]
Process Monitor show that conan.exe try run 
`C:\program "files" "(x86)\python35-32\python.exe" "C:\Program Files (x86)\Python35-32\Scripts\conan-script.py"`

So issue with python path that contain spaces

**Comment by memsharded** [comment]
Yes, certainly, thanks for debugging this. The only problem is that the conan.exe is not under our control, it is created by pip, so I guess very little we can do here. Maybe submit the issue to pip project?

**Comment by KindDragon** [comment]
I found pip issue https://github.com/pypa/pip/issues/2783
Please mention in this issue that it is important for your project

**Comment by lasote** [comment]
@KindDragon I see a merged PR and should be currently released in setuptools. I don't know if pypi needs to update something too. I've generated a pip package for the develop branch, Could someone try it?

pip install conan==0.12.1.dev0

**Comment by KindDragon** [comment]
Nope, same result: "failed to create process."

**Comment by lasote** [comment]
@KindDragon thanks, So pypi site have to update its setuptools too. :(

**Comment by wRastel** [comment]
So I had a similar problem yesterday, and it sounds like I can help with this issue. The windows installer for Python3+ 64-bit will put create the python directory under the Program Files (C:\Program Files\Python3x). When you try to run conan, you will be running the conan-script.py in the Python\Scripts directory. The first line of that file when using pip install and default windows installation of Python will look like this #!c:\program files\python35\python.exe. This is a problem because of the space in program files, but can be fixed by encapsulating the path in quotes like so that windows will treat the space as part of the path: #!"c:\program files\python35\python.exe". 

TL;DR Encapsulate the path in the first line of C:\Program Files\Python35\Scripts\conan-script.py to #!"c:\program files\python35\python.exe" to fix the issue

**Comment by Pixinn** [comment]
I also had the "_failed to create process._" message.

Adding quotes to the path of python in _conan-script.py_, as proposed by @wRastel solved this problem on my machine (Python 3.4 32bit, installed in "C:\Program Files (x86)\Python34\").

Maybe those quote should be added by default?

**Comment by memsharded** [comment]
Thanks @Pixinn for reporting. The thing is that it is out of our scope, and it is more a python issue than a conan issue. You can check https://github.com/conan-io/conan/blob/develop/setup.py, this is what we are using to create the python package. Is python/pip/setuptools what is creating that conan-script.py file with such path to your local python installation. 

https://github.com/pypa/pip/issues/2783
https://github.com/pypa/setuptools/issues/398

It seems that some solutions have ben provided, but maybe they are not merged in py3.4 and only in latest 3.5 or 3.6?

My suggestion: Install python in C:\Python3X, or maybe better, be clean, and install virtualenvwrapper-win and use virtualenvs for everything python related.

**Comment by Pixinn** [comment]
Thanks for your answer @memsharded. You're right, its a python issue and not a conan one, I did not realize it. Furthermore, this pb was indeed corrected in python 3.5.3.

**Comment by memsharded** [comment]
Thanks @Pixinn for telling. I think this issue can be closed now, nothing we can do here. I have added a note in the docs->Install instructions (will be published in a couple of days), clarifying this point.

**Comment by KindDragon** [comment]
JFYI issues still exist with Conan 0.21.2 and Python 3.5.2

**Comment by memsharded** [comment]
Hi @KindDragon , thanks for updating the status. Yes it is still failing ,but also still something that is out of our reach, it is under the control of Python. Unfortunately, we can't do anything to fix this, just recommend (as done in the notes of http://docs.conan.io/en/latest/installation.html) to install python in a path without spaces (or use virtualenvs-recommended for everything related to python, in some path without spaces)

**Comment by memsharded** [comment]
We have added a section in "troubleshoot", briefly explaining how to use virtualenvs, which is a practical and elegant solution to this issue: http://docs.conan.io/en/latest/faq/troubleshooting.html#error-failed-to-create-process

**Comment by bishwas-bh** [comment]
Yupp, this solution for this issues of [Unable to create process using in venv](https://webmatrices.com/d/503-unable-to-create-process-using-virtualenv). This will surely help you and fix your issues.

**Comment by MLIOSBond** [comment]
If anyone gets `failed to create process` or `Unable to create process` (after [creating a virtual environment][1]) &mdash; [this answer from SO][2] is on point.

I got these errors after installing VS 2022. The VS installer really does some weird things with spaces in paths...

I deleted VS logs in my user directory &mdash; and after that `conan` run just fine.

[1]: https://docs.conan.io/en/1.54/faq/troubleshooting.html#error-failed-to-create-process
[2]: https://stackoverflow.com/a/71054278

**Comment by memsharded** [comment]
Thanks for the tip @MLIOSBond. Certainly one of the things that I have learned to avoid is path with spaces in Windows, specially in user/account name. Really a pain for many tools out there.
